subsys/net/l2/wifi/wifi_utils.c

```diff
@@ -13,26 +13,202 @@ LOG_MODULE_REGISTER(net_wifi_utils, CONFIG_NET_L2_WIFI_MGMT_LOG_LEVEL);
+
 	if (!strcmp(band_str, "2")) {
-		return WIFI_FREQ_BAND_2_4_GHZ;
+		band = WIFI_FREQ_BAND_2_4_GHZ;
 	} else if (!strcmp(band_str, "5")) {
-		return WIFI_FREQ_BAND_5_GHZ;
+		band = WIFI_FREQ_BAND_5_GHZ;
 	} else if (!strcmp(band_str, "6")) {
-		return WIFI_FREQ_BAND_6_GHZ;
-	} else {
-		NET_ERR("Unknown band value: %s", band_str);
-		return WIFI_FREQ_BAND_UNKNOWN;
+		band = WIFI_FREQ_BAND_6_GHZ;
+	}
+
+	return band;
+}
+
+
+static bool wifi_utils_validate_chan_2g(uint16_t chan)
+{
+	if ((chan >= 1) && (chan <= 14)) {
+		return true;
+	}
+
+	return false;
+}
+
+
+static bool wifi_utils_validate_chan_5g(uint16_t chan)
+{
+	uint16_t i;
+
+	for (i = 0; i < ARRAY_SIZE(valid_5g_chans); i++) {
+		if (chan == valid_5g_chans[i]) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+
+static bool wifi_utils_validate_chan_6g(uint16_t chan)
+{
+	if (((chan >= 1) && (chan <= 233) && (!((chan - 1)%4))) ||
+	    (chan == 2)) {
+		return true;
+	}
+
+	return false;
+}
+
+
+static bool wifi_utils_validate_chan(uint8_t band,
+				     uint16_t chan)
+{
+	bool result = false;
+
+	switch (band) {
+	case WIFI_FREQ_BAND_2_4_GHZ:
+		result = wifi_utils_validate_chan_2g(chan);
+		break;
+	case WIFI_FREQ_BAND_5_GHZ:
+		result = wifi_utils_validate_chan_5g(chan);
+		break;
+	case WIFI_FREQ_BAND_6_GHZ:
+		result = wifi_utils_validate_chan_6g(chan);
+		break;
+	default:
+		NET_ERR("Unknown band: %d", band);
+		break;
+	}
+
+	return result;
+}
+
+
+static int wifi_utils_get_all_chans_in_range(uint16_t chan_start,
+		uint16_t chan_end,
+		uint16_t chan[__WIFI_FREQ_BAND_AFTER_LAST][WIFI_CHANNEL_MAX],
+		uint8_t band_idx,
+		uint8_t *chan_idx)
+{
+	uint16_t i;
+	bool start = false;
+	bool end = false;
+	uint8_t idx;
+
+	if (!wifi_utils_validate_chan(band_idx, chan_start)) {
+		NET_ERR("Invalid channel value %d in band %d", chan_start, band_idx);
+		return -EINVAL;
+	}
+
+	if (!wifi_utils_validate_chan(band_idx, chan_end)) {
+		NET_ERR("Invalid channel value %d in band %d", chan_end, band_idx);
+		return -EINVAL;
+	}
+
+	if (chan_end < chan_start) {
+		NET_ERR("Channel range end (%d) cannot be less than start (%d)",
+			chan_end,
+			chan_start);
+		return -EINVAL;
+	}
+
+	switch (band_idx) {
+	case WIFI_FREQ_BAND_2_4_GHZ:
+		idx = *chan_idx;
+
+		for (i = chan_start; i <= chan_end; i++) {
+			chan[band_idx][idx++] = i;
+		}
+
+		*chan_idx = idx;
+
+		break;
+	case WIFI_FREQ_BAND_5_GHZ:
+		idx = *chan_idx;
+
+		for (i = 0; i < ARRAY_SIZE(valid_5g_chans); i++) {
+			if (valid_5g_chans[i] == chan_start) {
+				start = true;
+			}
+
+			if (valid_5g_chans[i] == chan_end) {
+				end = true;
+			}
+
+			if (start) {
+				chan[band_idx][idx++] = valid_5g_chans[i];
+			}
+
+			if (end) {
+				*chan_idx = idx;
+				break;
+			}
+		}
+
+		break;
+	case WIFI_FREQ_BAND_6_GHZ:
+		idx = *chan_idx;
+
+		i = chan_start;
+
+		while (i <= chan_end) {
+			chan[band_idx][idx++] = i;
+
+			if (i == 1) {
+				i++;
+			} else if (i == 2) {
+				i += 2;
```

Comment:
IIUC, this becomes channel 4 which isn't valid in 6GHz right?

Reply:
Yeah thanks. Fixed this.